Adding support of SameSite cookies

IronBlood opened the pull request:
Linked issue #281

IronBlood commented:
Sorry to leave conflicts. I made this commit based on v0.8(46ed564), I thought git could handle the later changes automatically. Should I rebase to the master branch and create another PR?

davidmoreno commented on src/onion/response.h:
```diff
@@ -94,6 +94,9 @@ enum onion_response_flags_e{
 	OC_HTTP_ONLY=1, 				///< This cookie is not shown via javascript
 	OC_SECURE=2,						///< This cookie is sent only via https (info for the client, not the server).
+	OC_SAMESITE_NONE=0x10,				///< This cookie will be sent in all contexts, i.e. in responses to both first-party and cross-origin requests. When using this flag, OC_SECURE *MUST* also be used (or the cookie will be blocked by clients).
+	OC_SAMESITE_LAX=0x11,				///< This cookie won't be sent on normal cross-site subrequests (for example to load images or frames into a third party site), but are sent when a user is navigating to the origin site (i.e. when following a link). This is the default option to recent browser versions if none of these OC_SAMESITE_* flags are used.
```
I think this number for the flag will cause a conflict with OC_HTTP_ONLY, and same next line.

I would use 0x10, 0x20 and 0x40. I thought of combinations for 0x30 but actually they dont make sense, and if several OC_SAMESITE_* are marked its ok if it behaves incorrectly, as undefined behaviour. Altough a Warning can be written to the debug log.

davidmoreno commented on src/onion/response.c:
```diff
@@ -615,6 +615,20 @@ void onion_response_add_cookie(onion_response *res, const char *cookiename, cons
 	if (flags&OC_SECURE)
 		pos+=snprintf(data+pos, sizeof(data)-pos, "; Secure");
 
+	switch (flags & (OC_SAMESITE_NONE | OC_SAMESITE_LAX | OC_SAMESITE_STRICT)) {
```
Related to comment at onion_response_flags_e, I would use simple ifs as above as this switch does not really work as intended.

The simple ifs can be something like:

```c
if (flags&OC_SAMESITE_NONE) { 
... 
} else if (flags&OC_SAMESITE_LAX) {
  ...
} else if (flags&OC_SAMESITE_STRICT) {
 ...
}
```

And optionally some warning if several selected:

```c
if (check_if_several_bits_are_set...){
  WARNING("Must set only one of OC_SAMESITE_NONE, OC_SAMESITE_LAX or OC_SAMESITE_STRICT");
}
```

davidmoreno commented on tests/01-internal/03-response.c:
```diff
@@ -155,6 +155,33 @@ void t04_cookies(){
+	onion_dict_remove(h, "Set-Cookie");
+	onion_response_add_cookie(res, "key7", "value7", -1, NULL, NULL, OC_SAMESITE_NONE);
+	FAIL_IF_NOT_EQUAL_STR(onion_dict_get(h, "Set-Cookie"), "key7=value7; SameSite=None");
+
```
I miss in this tests also setting OC_HTTP_ONLY and OC_SECURE OR'ed with the OC_SAMESITE_*. 

Also the tests as of this version fail:

```
❯ ./03-response
03-response.c:39 INFO Test t01_create_add_free
[B63F1780] [2021-01-07 18:45:37] [ERROR response.c:270] Bad formed response. Need a request at creation. Will not write headers.
03-response.c:53 INFO t01_create_add_free ends: 3 succeses / 0 failures

03-response.c:59 INFO Test t02_full_cycle_http10
[B63F1780] [2021-01-07 18:45:37] [INFO response.c:186] [(null)] "GET (null)" 200 30 (Close connection)
03-response.c:92 INFO t02_full_cycle_http10 ends: 8 succeses / 0 failures

03-response.c:96 INFO Test t03_full_cycle_http11
[B63F1780] [2021-01-07 18:45:37] [INFO response.c:186] [(null)] "GET /" 200 30 (Keep-Alive)
03-response.c:133 INFO t03_full_cycle_http11 ends: 8 succeses / 0 failures

03-response.c:137 INFO Test t04_cookies
03-response.c:160 ERROR FAIL IF NOT EQUAL STR onion_dict_get(h, "Set-Cookie") != "key5=value5; path=/; domain=*.example.org; HttpOnly; Secure; SameSite=Strict" ("key5=value5; path=/; domain=*.example.org; HttpOnly; Secure" != "key5=value5; path=/; domain=*.example.org; HttpOnly; Secure; SameSite=Strict")
03-response.c:164 ERROR FAIL IF NOT EQUAL STR onion_dict_get(h, "Set-Cookie") != "key6=value6" ("key6=value6; HttpOnly; Secure" != "key6=value6")
03-response.c:167 ERROR FAIL IF NOT EQUAL STR onion_dict_get(h, "Set-Cookie") != "key6=value6" ("key6=value6; Secure; SameSite=Strict" != "key6=value6")
03-response.c:170 ERROR FAIL IF NOT EQUAL STR onion_dict_get(h, "Set-Cookie") != "key6=value6" ("key6=value6; HttpOnly; SameSite=Lax" != "key6=value6")
03-response.c:173 ERROR FAIL IF NOT EQUAL STR onion_dict_get(h, "Set-Cookie") != "key6=value6" ("key6=value6; HttpOnly; Secure" != "key6=value6")
03-response.c:177 ERROR FAIL IF NOT EQUAL STR onion_dict_get(h, "Set-Cookie") != "key7=value7; SameSite=Strict" ("key7=value7; Secure; SameSite=Strict" != "key7=value7; SameSite=Strict")
03-response.c:180 ERROR FAIL IF NOT EQUAL STR onion_dict_get(h, "Set-Cookie") != "key7=value7; SameSite=Lax" ("key7=value7; HttpOnly; SameSite=Lax" != "key7=value7; SameSite=Lax")
03-response.c:198 ERROR FAIL IF NOT valid_expires
[B63F1780] [2021-01-07 18:45:37] [ERROR response.c:270] Bad formed response. Need a request at creation. Will not write headers.
03-response.c:202 INFO t04_cookies ends: 6 succeses / 8 failures

03-response.c:206 INFO Test t05_printf
[B63F1780] [2021-01-07 18:45:37] [INFO response.c:186] [(null)] "GET (null)" 200 21 (Close connection)
03-response.c:228 INFO t05_printf ends: 1 succeses / 0 failures

03-response.c:232 INFO Test t06_empty
[B63F1780] [2021-01-07 18:45:37] [INFO response.c:186] [(null)] "GET (null)" 200 0 (Close connection)
03-response.c:258 INFO t06_empty ends: 2 succeses / 0 failures

03-response.c:262 INFO Test t07_large_printf
[B63F1780] [2021-01-07 18:45:37] [INFO response.c:186] [(null)] "GET (null)" 200 99999 (Close connection)
03-response.c:291 INFO t07_large_printf ends: 2 succeses / 0 failures

03-response.c:305 INFO TOTAL: 30 succeses / 8 failures


```

davidmoreno reviewed (commented):
Hi,

thanks a lot for the effort.

I detected some minor problems, mainly the chosen flag numbers, but it looks like soon we can merge it.

Thanks!

IronBlood commented:
Sorry for making troubles, I should have a second thought about picking flag numbers. I didn't know how to execute the tests back to that time, I remembered once there was a version that tests could be executed automatically after compilation.

After I rebased to the master branch and changed flags to `0x10` `0x20` and `0x40`, the tests looks good now (I also checked the returned value of `onion_response_add_cookie`) as shown below. I think it's better to close this PR and create a new one.

```
$ tests/01-internal/03-response                            
03-response.c:43 INFO Test t01_create_add_free
[DB7D2A00] [2021-01-08 12:53:58] [ERROR response.c:279] Bad formed response. Need a request at creation. Will not write headers.
03-response.c:57 INFO t01_create_add_free ends: 3 succeses / 0 failures

03-response.c:61 INFO Test t02_full_cycle_http10
[DB7D2A00] [2021-01-08 12:53:58] [INFO response.c:191] [(null)] "GET (null)" 200 30 (Close connection)
03-response.c:95 INFO t02_full_cycle_http10 ends: 8 succeses / 0 failures

03-response.c:99 INFO Test t03_full_cycle_http11
[DB7D2A00] [2021-01-08 12:53:58] [INFO response.c:191] [(null)] "GET /" 200 30 (Keep-Alive)
03-response.c:137 INFO t03_full_cycle_http11 ends: 8 succeses / 0 failures

03-response.c:141 INFO Test t04_cookies
03-response.c:229 ERROR FAIL IF NOT valid_expires
[DB7D2A00] [2021-01-08 12:53:58] [WARNING response.c:655] Cookie too long to be constructed. Not added to response.
[DB7D2A00] [2021-01-08 12:53:58] [ERROR response.c:279] Bad formed response. Need a request at creation. Will not write headers.
03-response.c:248 INFO t04_cookies ends: 27 succeses / 1 failures

03-response.c:252 INFO Test t05_printf
[DB7D2A00] [2021-01-08 12:53:58] [INFO response.c:191] [(null)] "GET (null)" 200 21 (Close connection)
03-response.c:275 INFO t05_printf ends: 1 succeses / 0 failures

03-response.c:279 INFO Test t06_empty
[DB7D2A00] [2021-01-08 12:53:58] [INFO response.c:191] [(null)] "GET (null)" 200 0 (Close connection)
03-response.c:306 INFO t06_empty ends: 2 succeses / 0 failures

03-response.c:310 INFO Test t07_large_printf
[DB7D2A00] [2021-01-08 12:53:58] [INFO response.c:191] [(null)] "GET (null)" 200 99999 (Close connection)
03-response.c:339 INFO t07_large_printf ends: 2 succeses / 0 failures

03-response.c:353 INFO TOTAL: 51 succeses / 1 failures
```

The warning and error of the `t04_cookies` part are caused by the test of "cookie too long" (I guess), I dumped the `data` after `ONION_WARNING` and find a lot of `a`s shown below:

```
cookie dump
key5=aaaaaaaaaaaaaa....
```

davidmoreno commented:
I think all this changes were also at PR #283, so I will close it now. If I'm wrong feel free to reopen the PR.
